### PR TITLE
Fix the TypeConverter when enabling UseSystemResourceKeys

### LINKED ISSUE #68651: CultureInfoConverter doesn't work correctly with UseSystemResourceKeys=true
There's a feature switch which is intended to reduce size of applications by removing exception messages and instead using just the resource key as the exception message. This is turned on by setting MSBuild property `UseSystemResourceKeys=true`.
The code which does that is here: https://github.com/dotnet/runtime/blob/4ec7a81108d944f292c7886653fc8f42494adc9d/src/libraries/Common/src/System/SR.cs#L21
It's compiled into each library and the rest of SR class (all the properties) are generated from resources.

This means that calling `SR.CultureInfoConverterDefaultCultureString` in default app returns `(Default)` from resources, but when `UseSystemResourceKeys=true` it returns `CultureInfoConverterDefaultCultureString` (string). This is used here: https://github.com/dotnet/runtime/blob/f10cddaf0863336327643f8c8e6ca332c596bed6/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/CultureInfoConverter.cs#L25

The external visible repro is:

`dotnet new console`

Program.cs:
```C#
System.ComponentModel.TypeDescriptor.GetConverter(typeof(System.Globalization.CultureInfo)).ConvertFrom(null, null, "(Default)");
```

`dotnet run` -> prints out nothing == works correctly.

`dotnet build /p:UseSystemResourceKeys=true`
`./bin/Debug/net6.0/app.exe` -> prints:
```
Unhandled exception. System.ArgumentException: CultureInfoConverterInvalidCulture, (Default) Arg_ParamName_Name, value
   at System.ComponentModel.CultureInfoConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value)
   at Program.<Main>$(String[] args) in ...\Program.cs:line 1
```

The `UseSystemResourceKeys` is set to true by default in Blazor WASM projects, so this is effectively broken in all Blazor projects by default.

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/68651

When enabling `UseSystemResourceKeys` msbuild property, the build will strip out all string resources and any resource lookup will report the resource key instead of the resource value. This is mainly used when reducing the size of the built binaries. In TypeConverter, this will cause a problem because there are some resources that have values with special meaning. Reporting the resource key in such cases will make the requested conversion operation fail as the converter will not recognize the value picked up from the resources. 

The fix is to apply @jkotas suggestion mentioned in the comment https://github.com/dotnet/runtime/issues/68651#issuecomment-1112641957.

### CODE REVIEW

**Comment by tarekgh** [comment]
@jkotas could you please have a look at this one? 

CC @eerhardt @vitek-karas

File: src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs [hunk 1]
```diff
@@ -322,7 +322,7 @@ public void Add2_Name_Duplicate()
             Assert.Equal(typeof(ArgumentException), ex.GetType());
             Assert.Null(ex.InnerException);
             Assert.NotNull(ex.Message);
-            Assert.True(ex.Message.IndexOf("'dup'") != -1);
+            Assert.True(ex.Message.IndexOf("'dup'") != -1 || ex.Message.IndexOf(" dup") != -1);
```
Comment (jkotas) [inline src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs:325]: What's the reason for omitting the apostrophes in the fallback string?

Comment (jkotas) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs:325]: Ah ok, this is the exception message. Makes sense.

**Review by jkotas — APPROVED** [review]
Thanks

File: src/libraries/System.ComponentModel.TypeConverter/tests/CultureInfoConverterTests.cs [hunk 2]
```diff
@@ -152,6 +153,15 @@ public void GetCultureName_Overriden_ConversionsReturnsExpected()
             Assert.Equal("Fixed", converter.ConvertTo(new CultureInfo("en-US"), typeof(string)));
         }
 
+        [Fact]
+        public void CultureInfoConverterForDefaultValue()
+        {
+            using (new ThreadCultureChange(null, CultureInfo.InvariantCulture))
+            {
+                Assert.Equal("", ((CultureInfo)TypeDescriptor.GetConverter(typeof(System.Globalization.CultureInfo)).ConvertFrom(null, null, "(Default)")).Name);
```
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/tests/CultureInfoConverterTests.cs:161]: Don't we need to run this test with `UseSystemResourceKeys=true` in order to test the bug?

File: src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs (hunk 1 above)
Comment (jkotas) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs:325]: The format string from resources is: `Duplicate component name '{0}' ...``

The synthetized format string that you get with UseSystemResourceKeys is `DuplicateComponentName {0}`. Notice that the apostrophes are missing.

**Comment by jkotas** [comment]
We should do a run of all libraries tests with `UseSystemResourceKeys=true`. I expect we will find more bugs like this one.

File: src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs:325]: OK, that makes sense. But I guess I don't understand how this test is getting run with `UseSystemResourceKeys=true`.

Comment (jkotas) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs:325]: I assume that @tarekgh has done a one-off manual run with `UseSystemResourceKeys=true`. I do not think we have it automated.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/Timers/TimersDescriptionAttribute.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ public override string Description
                 if (!_replaced)
                 {
                     _replaced = true;
-                    DescriptionValue = SR.Format(base.Description);
+                    DescriptionValue = string.Format(base.Description);
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/Timers/TimersDescriptionAttribute.cs:43]: You are right. I was keeping the old logic. for this one we don't have to have the format.

File: src/libraries/System.ComponentModel.TypeConverter/tests/CultureInfoConverterTests.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/System.ComponentModel.TypeConverter/tests/CultureInfoConverterTests.cs:161]: I have manually run the test with and without `UseSystemResourceKeys`. As @jkotas indicated, such mode should be tested for all libraries and not just this one only. I'll open infrastructure issue to do that.

File: src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs:325]: That is right, I have run the test with and without enabling UseSystemResourceKeys and ensured it is passing in both ways. Enabling `UseSystemResourceKeys` in general should be done separately from this PR. I'll open issue to track that.

Comment (tarekgh) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ContainerTests.cs:325]: I have logged the issue https://github.com/dotnet/runtime/issues/68707 to track enabling `UseSystemResourceKeys` on libraries tests.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.ComponentModel.TypeConverter/tests/CultureInfoConverterTests.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/tests/CultureInfoConverterTests.cs:161]: I'm not sure when we would enable a full test run with this setting. It doesn't seem like something that will happen anytime soon.

For now, we can run this one test with both `UseSystemResourceKeys=true or false` by doing something like the following code, only with the UseSystemResourceKeys switch:

https://github.com/dotnet/runtime/blob/66bd8b193bdaa96e26308e6c056f0b27434428cd/src/libraries/System.Diagnostics.Tracing/tests/BasicEventSourceTest/TestNotSupported.cs#L18-L30

Comment (tarekgh) [reply src/libraries/System.ComponentModel.TypeConverter/tests/CultureInfoConverterTests.cs:161]: ok, will do it.

**Comment by tarekgh** [comment]
@jkotas @stephentoub @eerhardt I need your feedback on an issue I am seeing in the test run regarding the change here. If we have the following code and running it without my changes:

```C#
            var attribute = new TimersDescriptionAttribute(null!); // passing null for description
            string s = null;
            try {s = attribute.Description; } catch (Exception e) { Console.WriteLine($"{e.Message}"); } // accessing attribute.Description first time throw argument null exception
            s = attribute.Description; // access it second time will not throw :-(
``` 

This is happening because the code https://github.com/dotnet/runtime/blob/8e624b33465546e9d05da2afea5faea233d9f847/src/libraries/System.ComponentModel.TypeConverter/src/System/Timers/TimersDescriptionAttribute.cs#L40. We are calling `SR.Format(string)` which will call [internal static string Format(string resourceFormat, params object?[]? args)](https://github.com/dotnet/runtime/blob/8e624b33465546e9d05da2afea5faea233d9f847/src/libraries/Common/src/System/SR.cs#L78) passing empty parameters and then later will throw when trying to format the string. 
With my changes, I removed the `SR.Format` as @eerhardt suggested and now we are not throwing it at all. Obviously, the current behavior is wrong. we have the following options:
- Throw on the attribute constructor when passing null constructor. This is the option I prefer. 
- Don't throw at all and allow `null` description. 
- Keep the current behavior which throw first time trying to access the description and never throw again.

What you think?

**Comment by vitek-karas** [comment]
I just realized that this fix only fixes half of the problem.
It will fix the bug as written, that is without trimming the code will work. But with trimming on, it will still break. That is because of this: https://github.com/dotnet/runtime/blob/8e624b33465546e9d05da2afea5faea233d9f847/eng/ILLink.Substitutions.Resources.template#L5

This instructs the linker to remove the resource string - all of them. Linker currently doesn't have the logic to figure out which resource strings are used by analyzing the code (it's actually quite difficult), so instead we work around it by simply removing all of the resources.

But this will only work correctly for libraries which only have exception messages in the resources.

So I think the fix for this library (and any other which has non-exception resources) should be to simply disable the "optimization". 
Basically follow https://github.com/dotnet/runtime/pull/42274 as per Eric's suggestion.

**Comment by jkotas** [comment]
> It will fix the bug as written, that is without trimming the code will work. But with trimming on, it will still break.

Why would it break? The library code does not load the strings from resources at all when you turn on `UseSystemResourceKeys`, so trimming the resources does not affect the behavior.

**Comment by jkotas** [comment]
> Throw in the attribute constructor when passing null parameter. This is the option I prefer.

This is a breaking change for no strong reason. I think keeping this behavior as-is would be fine. Or if you want to do something about this, the fix should be about removing exceptions thrown that is a lot less breaking.

**Comment by vitek-karas** [comment]
Sorry @jkotas you're right. This change adds the default values for all of the resources it needs. I didn't notice that and thought we go to the resource.

**Comment by tarekgh** [comment]
> This is a breaking change for no strong reason. I think keeping this behavior as-is would be fine. Or if you want to do something about this, the fix should be about removing exceptions thrown that is a lot less breaking.

With the current change is removing the exception throwing. but will have a risk of getting null description which can throw in other area in the app. If we are worried about the breaking, then we should keep the current behavior. I'll apply the fix to keep the current behavior then.
